src/devices/sound/hc55516.h

```diff
@@ -1,95 +1,212 @@
 // license:BSD-3-Clause
-// copyright-holders:Aaron Giles
+// copyright-holders:Aaron Giles, Jonathan Gevaryahu
+// thanks-to:Zonn Moore
 #ifndef MAME_SOUND_HC55516_H
 #define MAME_SOUND_HC55516_H
 
 #pragma once
 
-class hc55516_device : public device_t, public device_sound_interface
+class cvsd_device : public device_t, public device_sound_interface
 {
 public:
-	hc55516_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+	enum
+	{
+		RISING=true,
+		FALLING=false
+	};
+	cvsd_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	// common things across all the CVSD chips:
+	// Clock pull, really only relevant of something manually polls the clock (and clock is specified) for some reason, which is a very bad design pattern
+	// this function WILL ASSERT if it is called and the clock hz is NOT specified! TODO: remove this
+	READ_LINE_MEMBER( clock_r );
+
+	// Clock push
+	// this function WILL ASSERT if it is called and the clock hz IS specified!
+	WRITE_LINE_MEMBER( mclock_w );
+
+	// Digital in push to the pin, as a pseudo 'buffer'
+	WRITE_LINE_MEMBER( digin_w );
+
+	// DEC/ENC decode/encode select push
+	WRITE_LINE_MEMBER( dec_encq_w ); //TODO: handle encoding
+
+	// Digital out pull
+	READ_LINE_MEMBER( digout_r );
+
+	// callback versions of above functions
+	// A clock state change callback, because where we're going we don't need sanity and/or proper synchronization. TODO: remove this.
+	auto clock_state_cb() { return m_clock_state_push_cb.bind(); }
 
-	/* sets the digit (0 or 1) */
+	// Digital in pull callback function, for use if a clock is specified and we need to pull in the digital pin state, otherwise unused
+	//auto digin_cb() { return m_digin_pull_cb.bind(); } // TODO: this is not hooked up yet.
+
+	// Digital out push callback function.
+	//auto digout_cb() { return m_digout_push_cb.bind(); } // TODO: this is not hooked up yet.
```

Comment:
By convention, we group configuration members together near the top of the public section, before any live member functions.  It makes it more obvious what’s configuration and what isn't.  See pretty much any device in MAME:
* https://github.com/mamedev/mame/blob/mame0224/src/devices/machine/6525tpi.h#L46
* https://github.com/mamedev/mame/blob/mame0224/src/devices/machine/6532riot.h#L29
* https://github.com/mamedev/mame/blob/mame0224/src/devices/machine/at_keybc.h#L24
* https://github.com/mamedev/mame/blob/mame0224/src/devices/bus/nubus/nubus.h#L65

Reply:
I've changed the order for the configuration members so the callbacks appear first and grouped together, as in 6525tpi.h